packages/core/src/types.ts

```diff
@@ -75,6 +81,365 @@ export interface PositionWithOffset extends Position {
  */
 export type SourceRange = [number, number];
 
+//------------------------------------------------------------------------------
+// Rules
+//------------------------------------------------------------------------------
+
+/**
+ * What the rule is responsible for finding:
+ * - `problem` means the rule has noticed a potential error.
+ * - `suggestion` means the rule suggests an alternate or better approach.
+ * - `layout` means the rule is looking at spacing, indentation, etc.
+ */
+export type RuleType = "problem" | "suggestion" | "layout";
+
+/**
+ * The type of fix the rule can provide:
+ * - `code` means the rule can fix syntax.
+ * - `whitespace` means the rule can fix spacing and indentation.
+ */
+export type RuleFixType = "code" | "whitespace";
+
+/**
+ * An object containing visitor information for a rule. Each method is either the
+ * name of a node type or a selector, or is a method that will be called at specific
+ * times during the traversal.
+ */
+export interface RuleVisitor {
+	/**
+	 * Called for each node in the AST or at specific times during the traversal.
+	 */
+	[key: string]:
+		| ((node: unknown, parent?: unknown) => void)
+		| ((...unknown: unknown[]) => void);
+}
+
+/**
+ * Rule meta information used for documentation.
+ */
+export interface RulesMetaDocs {
+	/**
+	 * A short description of the rule.
+	 */
+	description?: string | undefined;
+
+	/**
+	 * The URL to the documentation for the rule.
+	 */
+	url?: string | undefined;
+
+	/**
+	 * The category the rule falls under.
+	 * @deprecated No longer used.
```

Comment:
> `@deprecated`

A thought: it would be nice if these types don't need to include deprecated things... is there a strong need to include them?

For example: if these are meant to be used in ESLint core, but ESLint core wouldn't include them until the next major version, maybe anything meant to be dropped in that next major version can be omitted?

Reply:
I don't have a strong opinion, it just seemed like it would be helpful to see something rather than not so that you know it wasn't an accidental omission in the type.

Reply:
>maybe anything meant to be dropped in that next major version can be omitted?

I feel deprecated types should be present, otherwise,  users using deprecated properties may run into errors. We can remove the types when removing the deprecated properties in the next major.